### ISSUE TITLE
[C#] Stream file uploads instead of copy all bytes into memory

### ISSUE THREAD

**Issue by hacki11** [swagger-api/swagger-codegen#2819]
ApiClient::ParameterToFile will copy complete file into ram before upload to rest service.

what about streaming a file to avoid OutOfMemory Exceptions on large files like:

```
public FileParameter ParameterToFile(string name, Stream stream)
        {
            var fp = new FileParameter
            {
                Writer = s =>
                {
                    using (StreamReader file = new StreamReader(stream))
                    {
                        file.BaseStream.CopyTo(s);
                    }
                },
                Name = name
            };


            if (stream is FileStream)
                fp.FileName = Path.GetFileName(((FileStream) stream).Name);
            else
                fp.FileName = "no_file_name_provided";
            return fp;
        }
```

current implementation:
[ApiClient.mustache](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/csharp/ApiClient.mustache#L214)
```
if (stream is FileStream)
                return FileParameter.Create(name, ReadAsBytes(stream), Path.GetFileName(((FileStream)stream).Name));
            else
                return FileParameter.Create(name, ReadAsBytes(stream), "no_file_name_provided");
```

**Comment by wing328** [comment]
@hacki11 thanks for the suggestion. Do you have cycle to submit a PR with the suggested change?

Ref: https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md

**Comment by hacki11** [comment]
Unfortunately, RestSharp 105.1.0 loads the complete body into memory before send.
The current RestSharp version has an issue not setting the Content-Length attribute where the rest server gets an eof exception....

**Comment by wing328** [comment]
@hacki11 are you referring to this issue? https://github.com/restsharp/RestSharp/issues/814

Roughly how large is your file?

**Comment by hacki11** [comment]
@wing328 i did not know there is already an issue - nice investigation!

it can easily grow to several hundred MB to 1 GB. 
Due to migrating my project to C# (java before) the problem arises. Java can stream perfectly with jaxrs.

**Comment by wing328** [comment]
@hacki11 you may want to try compiling the latest version (unofficial release) as another user has confirmed that it did address some of the issue with file upload: https://github.com/swagger-api/swagger-codegen/issues/2175

**Comment by hacki11** [comment]
i compiled the RestSharp master and the EOF Error on server side due to missing Content-Length is fixed now.
But the execute function of RestSharp seems to still load the whole stream into memory before streaming the content to the server.
There was already a stackoverflow discussion [SO](http://stackoverflow.com/questions/8988542/restsharp-is-loading-the-whole-file-into-memory-when-uploading-how-to-avoid-it#comment50800880_9005308) but it did not work for me

**Comment by wing328** [comment]
If you perform a debug, did you hit the following line mentioned in the SO comment?

https://github.com/restsharp/RestSharp/blob/62335a677c92ad45d9dc3214ff6765b597d8642f/RestSharp/Http.Sync.cs#L226

**Comment by hacki11** [comment]
yes, I hit that line and it uses the Action<Stream> of the FileParameter Writer Property.
After RestSharp ran the line above `file.BaseStream.CopyTo(s);` the process memory fills up with the amount of the file that is sent (tested with 250mb file).
i did assume that with this stream, the file will not be loaded completely into the ram.
i am not sure where the issue lies exactly.

**Comment by wing328** [comment]
@hacki11 I would suggest 2 things:

1) test with just RestSharp with a simple script to upload large files to see if you're experiencing the same so as so isolate the problem with RestSharp

2) upload a larger file (e.g. 2GB) to make sure you get the out of memory exception (so as to confirm the RestSharp is not smart enough to stream large files by not loading the whole file into memory)

**Comment by jimschubert** [comment]
It looks like RestSharp supports file streaming [via WriteMultipartFormData](https://github.com/restsharp/RestSharp/blob/105.1.0/RestSharp/Http.cs#L367). You'd have to set `AlwaysMultipartFormData = true` to hit this case:

```
RestRequest request = new RestRequest("/upload", Method.POST) 
{ 
  AlwaysMultipartFormData = true 
};
```

**Comment by wing328** [comment]
@hacki11 did you have a chance to review the feedback from @jimschubert and give it a try?

**Comment by hacki11** [comment]
it doesn't make a difference setting `AlwaysMultipartFormData = true` because the caller of the `WriteMultipartFormData` function checks the `HasFiles` flag which is true in my case:
```
Http.Sync.cs:
if (this.HasFiles || this.AlwaysMultipartFormData)
{
    this.WriteMultipartFormData(requestStream);
}
```
For completeness i set the AlwaysMultipartFormData = true, but the process consumed 1GB memory like the file to be uploaded.
![image](https://cloud.githubusercontent.com/assets/15319906/16429151/9f5245f6-3d75-11e6-8173-de0ca64b9a9d.png)

Debugging makes me wonder where the issue lies.
WriteMultipartFormData tries to write the stream to the webrequest stream (seems correct to me):
![image](https://cloud.githubusercontent.com/assets/15319906/16429319/5f961e64-3d76-11e6-96f7-fa2f1054131f.png)

This function calls my Writer Action in the FileParameter object:
![image](https://cloud.githubusercontent.com/assets/15319906/16429345/8de03822-3d76-11e6-9fa5-ed54967393aa.png)
After executing the CopyTo function memory fills up

---- 

After some tests i end up with a OutOfMemory Exception already at this line `using (Stream requestStream = webRequest.GetRequestStream())` but can't tell really why.

**Comment by jimschubert** [comment]
@hacki11 thanks for your investigation work. 

Looks like you'd get the exception on that line because ParameterToFile is flushing the stream.

Do you have a test project repo you used for investigation, or were you testing against your codebase?

**Comment by chenzhekl** [comment]
Seems to be related to this https://github.com/restsharp/RestSharp/issues/889

**Comment by wing328** [comment]
@chenzhekl thanks for the pointer. You may be interested in this task as well: https://github.com/swagger-api/swagger-codegen/issues/7012

**Comment by nevergiveupc** [comment]
Is this problem solved or not? @hacki11 @wing328

**Comment by alexeyzimarev** [comment]
Don't use `FileParameter`. Use `AddFile` overload that accepts the stream writer delegate.
